server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -87,7 +87,9 @@ private static FieldFetcher create(
                 if (nestedParentPaths.isEmpty() == false) {
                     // try to find the shortest nested parent path for this field
                     for (String nestedFieldPath : nestedParentPaths) {
-                        if (field.startsWith(nestedFieldPath)) {
+                        if (field.startsWith(nestedFieldPath)
+                            && field.length() > nestedFieldPath.length()
+                            && field.charAt(nestedFieldPath.length()) == '.') {
```

Comment:
Looks good to me, maybe @romseygeek has some comments on directly checking the field separator though because he is currently working on allowing dots in fields names. As far as I understand this should be okay here though, just checking for making sure...

Reply:
This is fine for now. I have a PR which simplifies some of this nested path handling which I think will help here as well in the longer term (#75677) but the newly added test should ensure that things work correctly.